### DISCUSSION TITLE
Teaching Rust to Python developers

### DISCUSSION THREAD

**Discussion by tomshlomo** [google/comprehensive-rust#2502] [Show and tell]
Recently, at my workplace (Via), a colleague and I taught the Comprehensive Rust course to a group of about 15 Python developers.

The experience was fantastic—both we and the students learned a great deal and had a lot of fun. A big thank you for creating such an amazing resource!

We made a few modifications, primarily replacing some of the exercises and adding a section on using PyO3. We've published everything [here](https://tomshlomo.github.io/rust-with-via/index.html).

We hope our additions are useful to others exploring Rust and would appreciate any feedback!

**Comment by djmitche** [comment]
That non-empty vec, in particular, is a neat idea. Where did you insert these exercises into the course?

**Reply by tomshlomo** [reply]
The non-empty vec exercise replaced the health statistics exercise.
For more information you can check out our [schedule](https://tomshlomo.github.io/rust-with-via/schedule.html) page.

**Reply by mgeisler** [reply]
> The non-empty vec exercise replaced the health statistics exercise.

That exercise is probably the one I like the least :sweat_smile: I would very happy to replace it in our repository too :smile: 

@randomPoison and @fw-immunant (our two main instructors at Google), please let me know if you have any concerns!

**Reply by randomPoison** [reply]
Yeah I often skip the health statistic exercise in the interest of time so that students have as much time as possible for the protobuf exercise (which covers the same topics but much better imo). I very much like this non-empty vec exercise, but it seems like a longer one and so might not fit in the same day as the protobuf exercise.

**Reply by tomshlomo** [reply]
@randomPoison we actually also replaced the protobuf exercise with [this one](https://tomshlomo.github.io/rust-with-via/lifetimes/index.html) (which is based on [this one](https://github.com/tfpk/lifetimekata/blob/main/exercises/05_lifetimes_on_types/exercise/src/lib.rs) from [LifetimeKata](https://tfpk.github.io/lifetimekata)).

The motivation for this change is that, on our team, tasks like parsing and strings/bytes manipulation are relatively uncommon.

Additionally, I think seeing a concrete example where using a single lifetime specifier instead of two results in a compilation error is really helpful for understanding lifetime annotations. The exercise demonstrates this by allowing most tests to pass with a single lifetime specifier, while one test can only pass with two lifetime specifiers.

**Reply by randomPoison** [reply]
Oh that's great. As much as I like the protobuf exercise, it is pretty complicated and I think students spend a lot of time just wrapping their heads around what they need to do. This seems way more concise and understandable, and better demonstrates the usage of lifetimes.

I think both of these are nice improvements over our current ones, so I think these would be great to integrate :)

**Reply by tomshlomo** [reply]
@randomPoison @mgeisler I’d be happy to contribute! What’s the best way to proceed? Should I just open a PR to replace the exercise(s)? That feels a bit too drastic :)

**Reply by mgeisler** [reply]
Hi @tomshlomo, actually, a PR would be great! I'm sorry that I never replied on this, but @djmitche just reminded me of this thread.

**Reply by djmitche** [reply]
I'm happy to help with the PR, too, but your authorship will ensure attribution and licensing and all of that are OK.

**Comment by mgeisler** [comment]
Thanks for sharing this! It's super nice to hear that the course is useful around the world!
